### PR TITLE
Adjust cancel behavior around remaining tickets

### PR DESCRIPTION
**Pull request by Timo614**
### Closes Issues:
Closes: #881

### Description:

This branch fixes the associated issue by marking unused tickets as nullified similar to the behavior that occurs when you update a `ticket_type` to have less quantity than it currently has. I adjusted the reporting logic to exclude the ticket types with no non nullified tickets as well.

Ticket types with no sold orders fall off the dashboard page once cancelled (aside from the details where they are marked as cancelled but do not appear in the listing, capacity, or reports). Ones that had sold tickets have only their unsold marked as `Nullified` and the other ones still appear in the reports and dashboard numbers.

May be worth having someone try the branch with their copy of bn-web and confirm I didn't miss any other places this may appear. I spent a little while looking so I think I covered all of the areas in the tooling this shows up though.

Open questions (mentioned in associated issue but persisting here):
- How do we want the logic to behave in the event some of these tickets are in the cart at the time of cancelling? They would become available again as soon as the reservation ended so they could end up back in inventory again. Right now I'm not doing anything with them (hence they appear in the `Open` number below).
- For the partial payments logic where a partial payment can reserve tickets for up to a day would we want to trigger a refund on any partially paid orders if we plan on removing them above? Wondering how the cart would show this to the user.

I noticed during this branch / maybe worth adjusting in other issues but:
- Holds can have their set quantity lowered even below the number of units currently reserved in the cart
- Holds do not currently respect the `end_at` value they have set
- Hold purchases appear in both the `Sold` and `Held` values in the dashboard which may be misleading since I'd think the sold should only be in the former once they have been purchased
- The revenue numbers appear off by $0.10 in the photo below

screen shot 2019-01-17 at 4 30 09 pm

You can see the latter here. There are 90 tickets left after 30 held and 60 regular. The 30 held are all purchased already. 11 of the regular are in a sold state and 49 are in a reserved state. The reserved state are shown as `Open` here. It looks a bit confusing on first glance without calculating those out.

### Migrations
 * No Migrations

### Environment Variables
 * No change

### CODE REVIEW

**Comment by mikethetike** [comment]
@zmelamed please have a look at @Timo614 's comments regarding the holds

File: api/src/controllers/ticket_types.rs [hunk 1]
```diff
@@ -167,17 +168,38 @@ pub fn index(
...
     let ticket_type = TicketType::find(path.ticket_type_id, connection)?;
-
     ticket_type.cancel(connection)?;
 
+    // Reduce holds to quantity sold
+    for hold in Hold::find_by_ticket_type(ticket_type.id, connection)?
+        .iter()
+        .filter(|h| h.parent_hold_id.is_none())
+    {
+        hold.remove_available_quantity(connection)?;
```
Comment (mikethetike) [inline api/src/controllers/ticket_types.rs]: A hold can only belong to one ticket type, so we should delete the hold once it's empty

File: db/src/models/ticket_types.rs [hunk 2]
```diff
@@ -269,6 +269,21 @@ impl TicketType {
         Ok(valid_ticket_count as u32)
     }
 
+    pub fn valid_unsold_ticket_count(&self, conn: &PgConnection) -> Result<u32, DatabaseError> {
+        let valid_unsold_ticket_count: i64 = ticket_instances::table
+            .inner_join(assets::table)
+            .filter(assets::ticket_type_id.eq(self.id))
+            .filter(ticket_instances::status.ne(TicketInstanceStatus::Nullified))
```
Comment (mikethetike) [inline db/src/models/ticket_types.rs]: The next statement makes this line redundant. Did you intend a different column?

File: api/src/controllers/ticket_types.rs [hunk 3]
```diff
@@ -167,17 +168,38 @@ pub fn index(
...
+
+    let valid_unsold_ticket_count = ticket_type.valid_unsold_ticket_count(connection)?;
+    nullify_tickets(
+        state,
+        organization,
+        ticket_type,
+        valid_unsold_ticket_count,
+        connection,
+    )?;
+
```
Comment (mikethetike) [inline api/src/controllers/ticket_types.rs]: Should we not add a check to see that all of the ticket instances are now nullified?

File: api/src/controllers/ticket_types.rs (hunk 1 above)
Comment (Timo614) [reply api/src/controllers/ticket_types.rs]: Sounds good, will do that if there are no sold tickets for the hold.

File: api/src/controllers/ticket_types.rs (hunk 3 above)
Comment (Timo614) [reply api/src/controllers/ticket_types.rs]: We kind of do that already via the `nullify_tickets` method.

https://github.com/big-neon/bn-api/pull/888/files#diff-3cdd11b8c61ac0e1afaa3fa5020e1339R377

If the quantity we requested to be nullified does not match the quantity that was actually nullified the logic errors out reverting the change. We'd still want some to be in the `Purchased` and `Redeemed` states so the logic would be just confirming that the new `valid_unsold_ticket_count` was 0 so it effectively would work the same way as that check.

File: db/src/models/ticket_types.rs (hunk 2 above)
Comment (Timo614) [reply db/src/models/ticket_types.rs]: Nope, was a mistake as I iterated on this. Will remove the ne nullified line here.

**Comment by Timo614** [comment]
> I think for tickets that are currently reserved in someone's cart, they should be cancelled. We must then add a check to see that all of the items in the cart 1. are still reserved (if we don't do this check already) and 2. have not been cancelled. If either are not true, the checkout should fail.

Sounds good for this card I'll modify the logic to also nullify reserved tickets and will make a note to make a follow up issue to check if a ticket is still reserved when in the cart and checking out. We may want to also modify the display cart logic to signal if there's a problem so a user can be alerted prior to payment so will make a note in that follow up issue for that.

File: api/src/controllers/ticket_types.rs (hunk 3 above)
Comment (mikethetike) [reply api/src/controllers/ticket_types.rs]: Ok cool

**Pull request merged**